mlrun/model_monitoring/applications/context.py

```diff
@@ -49,79 +56,89 @@ class MonitoringApplicationContext(MLClientCtx):
     :param model_endpoint:          (ModelEndpoint) The model endpoint object.
     :param feature_names:           (list[str]) List of models feature names.
     :param label_names:             (list[str]) List of models label names.
-    :param model:                   (tuple[str, ModelArtifact, dict]) The model file, model spec object, and list of
-
+    :param model:                   (tuple[str, ModelArtifact, dict]) The model file, model spec object,
+                                    and list of of extra data items.
     """
 
-    def __init__(self, **kwargs):
-        super().__init__(**kwargs)
-
-    def _enrich_data(self):
-        self.application_name: typing.Optional[str] = None
-        self.start_infer_time: typing.Optional[pd.Timestamp] = None
-        self.end_infer_time: typing.Optional[pd.Timestamp] = None
-        self.latest_request: typing.Optional[pd.Timestamp] = None
-        self.endpoint_id: typing.Optional[str] = None
-        self.output_stream_uri: typing.Optional[str] = None
-
-        self._sample_df: typing.Optional[pd.DataFrame] = None
-        self._model_endpoint: typing.Optional[ModelEndpoint] = None
-        self._feature_stats: typing.Optional[FeatureStats] = None
-        self._sample_df_stats: typing.Optional[FeatureStats] = None
-
-    @classmethod
-    def from_dict(
-        cls,
-        attrs: dict,
-        context=None,
-        model_endpoint_dict=None,
-        **kwargs,
-    ) -> "MonitoringApplicationContext":
+    def __init__(
+        self,
+        *,
+        graph_context: mlrun.serving.GraphContext,
+        application_name: str,
+        event: dict[str, Any],
+        model_endpoint_dict: dict[str, ModelEndpoint],
+    ) -> None:
         """
-        Create an instance of the MonitoringApplicationContext from a dictionary.
+        Initialize a `MonitoringApplicationContext` object.
+        Note: this object should not be instantiated manually.
 
-        :param attrs:               The instance data dictionary.
-        :param context:             The current application context.
+        :param application_name:    The application name.
+        :param event:               The instance data dictionary.
         :param model_endpoint_dict: Dictionary of model endpoints.
-
         """
+        self.application_name = application_name
 
-        if not context:
-            ctx = (
-                super().from_dict(
-                    attrs=attrs.get(mm_constants.ApplicationEvent.MLRUN_CONTEXT, {}),
-                    **kwargs,
-                ),
-            )
-        else:
-            ctx = context
-            cls._enrich_data(ctx)
+        self.project_name = cast(str, mlrun.mlconf.default_project)
```

Comment:
The default project is overridden in a Nuclio runtime with the appropriate environment variable, e.g.:
![image](https://github.com/user-attachments/assets/bdfabe0e-0f93-4a07-8295-f183cd3a8e42)

The artifacts logging is checked at the project level by unit tests and system tests, and they all passed 🙂

Do you prefer taking it from the graph context? After checking its implementation, I'm not sure it's cleaner.

Reply:
To be honest, I wasn't aware of this - now that I know it's there I can see where it's happening in the code, and I see why it works. However, this can be very confusing, assuming you're not aware of this little behind-the-scenes magic MLRun is doing, and reading the code can be misleading. 
Also, once we enable debug-mode for MM apps and allow running them in some job or locally (see [ML-6362](https://iguazio.atlassian.net/browse/ML-6362)) - this will stop working, since this enrichment is unique to nuclio functions.
 
So, while for now it works I think it may be better to not rely on this. And I tend to agree that the graph-context implementation also seems a bit shaky, but I think it works across the board.
Another option is to keep this, but add a proper comment there to explain why this works *for nuclio* and also a TODO to fix once debug mode is implemented.